### PR TITLE
Use dynamic add actions for topology context menu

### PR DESCRIPTION
**Pull request by rottencandy**
**Fixes**: 
https://issues.redhat.com/browse/ODC-6349

**Description**: 
Use extension based Add actions instead of hard coded actions for topology context menu of add to project.

**Screen shots / Gifs for design review**: 

**Unit test coverage report**: 

**Browser conformance**: 

- [ ] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

/kind feature

### CODE REVIEW

File: frontend/packages/topology/src/actions/context-menu.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import { INCONTEXT_ACTIONS_CONNECTS_TO, QUERY_PROPERTIES } from '@console/dev-console/src/const';
+import { resolvedHref } from '@console/dev-console/src/utils/add-page-utils';
```
Comment (sahil143) [inline frontend/packages/topology/src/actions/context-menu.ts:2]: These are bad dependencies

File: frontend/packages/topology/src/components/application-panel/TopologyApplicationPanel.tsx [hunk 2]
```diff
@@ -1,32 +1,42 @@
...
+      <div className="overview__sidebar-pane-head resource-overview__heading">
+        <h1 className="co-m-pane__heading">
+          <div className="co-m-pane__name co-resource-item">
+            <ResourceIcon className="co-m-resource-icon--lg" kind="application" />
+            {application.name}
+          </div>
+          <div className="co-actions">
+            <ActionServiceProvider context={{ 'topology-context-actions': element }}>
+              {({ options, loaded }) =>
+                loaded && <ActionsMenu actions={groupActions(options, graphData, application)} />
```
Comment (sahil143) [inline frontend/packages/topology/src/components/application-panel/TopologyApplicationPanel.tsx:31]: we shouldn't be calling groupActions here to get the delete action. It should come from 
ActionServiceProvider

File: frontend/packages/topology/src/actions/provider.ts [hunk 3]
```diff
@@ -17,3 +22,39 @@ export const useTopologyWorloadActionProvider = (element: GraphElement) => {
...
+  const loaded =
+    !allAddActionsDisabled && addActionExtensionsResolved && filteredAddActionExtensionsLoaded;
+
+  const loadedActions = useMemo(
+    () => loaded && filteredAddActionExtensions.filter((a) => !!a.properties.action),
+    [loaded, filteredAddActionExtensions],
+  );
+
+  return useMemo(() => {
+    if (element.getKind() === ModelKind.graph) {
```
Comment (sahil143) [inline frontend/packages/topology/src/actions/provider.ts:49]: create separate action provider for graph and application

File: frontend/packages/topology/src/actions/provider.ts [hunk 4]
```diff
@@ -17,3 +22,39 @@ export const useTopologyWorloadActionProvider = (element: GraphElement) => {
...
+      const actions = loaded && loadedActions.map((a) => contextAction(a, namespace));
+      return [actions, loaded, undefined];
+    }
+    if (element.getType() === TYPE_APPLICATION_GROUP) {
```
Comment (sahil143) [inline frontend/packages/topology/src/actions/provider.ts:53]: you can add deleteAction in an array with other application actions if you create a separate action provider

File: frontend/packages/topology/src/actions/provider.ts [hunk 5]
```diff
@@ -1,8 +1,13 @@
 import { useMemo } from 'react';
-import { GraphElement } from '@patternfly/react-topology';
+import { GraphElement, ModelKind } from '@patternfly/react-topology';
+import { useAccessFilterExtensions } from '@console/dev-console/src/components/add/hooks/useAccessFilterExtensions';
+import { filterNamespaceScopedUrl } from '@console/dev-console/src/utils/add-page-utils';
+import { useAddActionExtensions } from '@console/dev-console/src/utils/useAddActionExtensions';
```
Comment (sahil143) [inline frontend/packages/topology/src/actions/provider.ts:5]: These are bad dependencies

File: frontend/packages/topology/src/actions/provider.ts [hunk 6]
```diff
@@ -17,3 +22,39 @@ export const useTopologyWorloadActionProvider = (element: GraphElement) => {
     return [actions, true, undefined];
   }, [actions, element]);
 };
+
+export const useTopologyContextActionProvider = (element: GraphElement) => {
```
Comment (sahil143) [inline frontend/packages/topology/src/actions/provider.ts:26]: I would suggest moving these action providers to dev-console package.

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/topology/src/actions/provider.ts (hunk 6 above)
Comment (rottencandy) [reply frontend/packages/topology/src/actions/provider.ts]: Moved providers to dev console package.

File: frontend/packages/topology/src/actions/provider.ts (hunk 3 above)
Comment (rottencandy) [reply frontend/packages/topology/src/actions/provider.ts]: Updated.

**Comment by debsmita1** [comment]
Cannot see Helm charts , upload jar file , database options
Screenshot from 2021-09-24 00-42-13

File: frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx [hunk 7]
```diff
@@ -76,17 +76,23 @@ export const noRegroupWorkloadContextMenu = (element: Node) =>
...
+      }
+    </ActionServiceProvider>,
+  ];
 };
 
-export const graphContextMenu = (graph: Graph, connectorSource?: Node) =>
-  createMenuItems(kebabOptionsToMenu(graphActions(graph.getData(), connectorSource)));
+export const graphContextMenu = (graph: Graph) => {
+  return [
+    <ActionServiceProvider key="topology" context={{ 'topology-context-actions': graph }}>
```
Comment (debsmita1) [inline frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx:92]: any reason why we can't have `topology-actions` as the context id ?

File: frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx [hunk 8]
```diff
@@ -76,17 +76,23 @@ export const noRegroupWorkloadContextMenu = (element: Node) =>
...
+      {({ options, loaded }) =>
+        loaded && createMenuItems(kebabOptionsToMenu(graphActions(options)))
```
Comment (debsmita1) [inline frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx:94]: Can you try using `createContextMenuItems` ?

File: frontend/packages/dev-console/src/actions/context-menu.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+export const applicationDeleteAction = (application: TopologyApplicationObject): Action => {
+  const primaryResource = application.resources[0].resource;
+  const resourceModel = modelFor(primaryResource.kind)
+    ? modelFor(primaryResource.kind)
+    : modelFor(referenceFor(primaryResource));
+  return {
+    id: 'delete-application',
+    // t('devconsole~Delete Application')
+    label: 'devconsole~Delete Application',
```
[the remark this answers is not in the document]
Comment (rottencandy) [reply frontend/packages/dev-console/src/actions/context-menu.ts]: Updated.

File: frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx (hunk 7 above)
Comment (rottencandy) [reply frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx]: The provider for `topology-actions` context always expects the node to have a single associated resource in the element node, and crashes if the graph/application node is provided since they have multiple resources.

File: frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx (hunk 8 above)
Comment (rottencandy) [reply frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx]: Updated. Thanks!

**Comment by rottencandy** [comment]
> Cannot see Helm charts , upload jar file , database options

Added upload jar file & serverless add options, I haven't added helm & database because they were not shown previously.

**Comment by debsmita1** [comment]
> > Cannot see Helm charts , upload jar file , database options
> 
> Added upload jar file & serverless add options, I haven't added helm & database because they were not shown previously.
> 

yes the knative add options were not showing up , I wrote jar file option instead . My bad.

**Comment by debsmita1** [comment]
getting `Maximum update depth exceeded` warnings

**Comment by rottencandy** [comment]
> getting `Maximum update depth exceeded` warnings

Was happening because `useAccessFilterExtensions` was not fully memoized. Fixed.

File: frontend/packages/dev-console/src/actions/providers.ts [hunk 10]
```diff
@@ -16,3 +24,71 @@ export const useEditImportActionProvider = (resource: K8sResourceKind) => {
...
+  const loaded =
+    !allAddActionsDisabled && addActionExtensionsResolved && filteredAddActionExtensionsLoaded;
+
+  return React.useMemo(() => {
+    if (element.getKind() === ModelKind.graph) {
+      const actions =
+        loaded &&
+        filteredAddActionExtensions
+          .filter((a) => !!a.properties.action)
+          .map((a) => contextAction(a, namespace));
```
Comment (sahil143) [inline frontend/packages/dev-console/src/actions/providers.ts:51]: pass the path here to contextAction

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts [hunk 11]
```diff
@@ -19,6 +19,11 @@ export type AddAction = ExtensionDeclaration<
     icon?: CodeRef<React.ReactNode>;
     /** Optional access review to control visibility / enablement of the action. */
     accessReview?: AccessReviewResourceAttributes[];
+    /** action to specify add action as a topology context menu item */
+    action?: {
+      name?: string;
+      isGroupable: boolean;
```
Comment (sahil143) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts:25]: add a comment here noting.. `isGroupable` is to denote that resources created through this flow are groupable or not

File: frontend/packages/dev-console/src/actions/providers.ts [hunk 12]
```diff
@@ -16,3 +24,71 @@ export const useEditImportActionProvider = (resource: K8sResourceKind) => {
...
+        id: element.getId(),
+        name: appName,
+        resources: element.getData().groupResources,
+      };
+
+      const actions = loaded && [
+        ...filteredAddActionExtensions
+          .filter((a) => a.properties.action?.isGroupable)
+          .map((a) => contextAction(a, namespace, appName)),
+        applicationDeleteAction(appData),
```
Comment (sahil143) [inline frontend/packages/dev-console/src/actions/providers.ts:88]: delete action should come before all other menu items as it is on master

image

**Review by sahil143 — CHANGES_REQUESTED** [review]

**Review by sahil143 — CHANGES_REQUESTED** [review]
connector source is not working properly

image
Master:
image

connector source on graph:
image

Master:
image

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts [hunk 13]
```diff
@@ -19,6 +19,12 @@ export type AddAction = ExtensionDeclaration<
     icon?: CodeRef<React.ReactNode>;
     /** Optional access review to control visibility / enablement of the action. */
     accessReview?: AccessReviewResourceAttributes[];
+    /** action to specify add action as a topology context menu item */
+    action?: {
```
[the remark this answers is not in the document]
Comment (rottencandy) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts]: Switched back to using hardcoded actions.

**Comment by rottencandy** [comment]
> connector source is not working properly

Fixed.

**Comment by debsmita1** [comment]
got an error when clicked on(left-click) application group
app-error

File: frontend/packages/dev-console/src/actions/add-resources.tsx [hunk 14]
```diff
@@ -8,125 +8,138 @@ import {
...
+export const AddActions: { [name: string]: ActionFactory } = {
+  FromGit: (namespace, application, contextSource, path, accessReviewDisabled) => ({
+    id: 'import-from-git',
+    label: i18next.t('devconsole~Import from Git'),
+    icon: <GitAltIcon />,
+    cta: {
+      href: resolvedURLWithParams('/import/ns/:namespace', namespace, application, contextSource),
+    },
+    path,
+    disabled: !accessReviewDisabled,
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/actions/add-resources.tsx:60]:
```suggestion
    disabled: accessReviewDisabled,
```

do the same for the other add actions

File: frontend/packages/dev-console/src/actions/providers.ts [hunk 15]
```diff
@@ -56,19 +63,107 @@ const resourceAttributes = (model: K8sKind, namespace: string): AccessReviewReso
...
+    secretAccess &&
+    routeAccess &&
+    serviceAccess;
+  const isCatalogImageResourceAccess = isImportResourceAccess && imageStreamImportAccess;
+
+  return React.useMemo(() => {
+    if (element.getKind() === ModelKind.graph) {
+      const actions = connectorSource
+        ? [
+            AddActions.FromGit(namespace, undefined, undefined, '', isImportResourceAccess),
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/actions/providers.ts:95]:
```suggestion
            AddActions.FromGit(namespace, undefined, undefined, '', !isImportResourceAccess),
```

**Review by debsmita1 — CHANGES_REQUESTED** [review]

**Comment by debsmita1** [comment]
Event source option is missing
Before
before

After
after

**Comment by rottencandy** [comment]
Updated to remove edge connector for KafkaSource on drag end and use `ActionFilter`.

https://user-images.githubusercontent.com/20013884/139873517-2105e140-30ba-4d85-a153-a31c402e0438.mp4

File: frontend/packages/topology/src/actions/TopologyActions.tsx [hunk 16]
```diff
@@ -13,7 +13,7 @@ const TopologyActions: React.FC<TopologyActionsProps> = ({ element }) => {
     const resource = getResource(element);
     const { csvName } = element.getData()?.data ?? {};
     return {
-      'topology-actions': element,
+      'topology-context-actions': { element },
```
Comment (sahil143) [inline frontend/packages/topology/src/actions/TopologyActions.tsx:16]: Why did you remove `topology-actions`?

Comment (rottencandy) [reply frontend/packages/topology/src/actions/TopologyActions.tsx:16]: Changed to `topology-context-actions` because many of the add to application actions in `topology-actions` were not coming from dynamic plugins, but hardcoded in add-resources-menu-utils.ts which I removed in this PR.

Comment (sahil143) [reply frontend/packages/topology/src/actions/TopologyActions.tsx:16]: we use `topology-actions` for helm actions as well. So this will break that?

Comment (rottencandy) [reply frontend/packages/topology/src/actions/TopologyActions.tsx:16]: Added helm actions to helm's topology provider.

Comment (sahil143) [reply frontend/packages/topology/src/actions/TopologyActions.tsx:16]: it's not just helm topolofy-actions context is being used in other places as well and the provider doesn't expect `{ element }` in the argument.
Please add `topology-context-actions` as a seperate context

Comment (rottencandy) [reply frontend/packages/topology/src/actions/TopologyActions.tsx:16]: Oh, did not know that a single consumer could use multiple contexts : p updated.

**Review by sahil143 — APPROVED** [review]
/lgtm

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
Overall changes looks good. I found a few issues while running the PR on a cluster - 

- The in-context connection doesn't work anymore. The resource is created but the connection is not created.
- The UI breaks when I click on the explicitly created connector between two resources.

https://user-images.githubusercontent.com/6041994/144271949-af3a7dcd-cf0e-4f39-8859-e245483d4b8d.mov

**Comment by rottencandy** [comment]
> @rottencandy Overall changes looks good. I found a few issues while running the PR on a cluster -
> 
>     * The in-context connection doesn't work anymore. The resource is created but the connection is not created.
> 
>     * The UI breaks when I click on the explicitly created connector between two resources.

Fixed in-context connection creation & connector sidebar crash.

**Comment by rottencandy** [comment]
Propagating Docs & PX ack labels from epic [ODC-6351](https://issues.redhat.com/browse/ODC-6351)
/label docs-approved
/label px-approved

**Review by rohitkrai03 — APPROVED** [review]
/lgtm
/approve
